src/api/app/models/worker_status.rb

```diff
@@ -44,88 +44,83 @@ def project_names(prjs)
 
   def update_workerstatus_cache
     # do not add hiding in here - this is purely for statistics
-    ret = Backend::Api::BuildResults::Worker.status
-    wdata = Xmlhash.parse(ret)
+    backend_status = Backend::Api::BuildResults::Worker.status
+    wdata = Nokogiri::XML(backend_status)
     @mytime = Time.now.to_i
-    @squeues = {}
-    Rails.cache.write('workerstatus', ret, expires_in: 3.minutes)
+    @squeues = Hash.new(0)
+
+    Rails.cache.write('workerstatus', backend_status, expires_in: 3.minutes)
+
     StatusHistory.transaction do
-      wdata.elements('blocked') do |e|
-        save_value_line(e, 'blocked')
-      end
-      wdata.elements('waiting') do |e|
-        save_value_line(e, 'waiting')
-      end
-      wdata.elements('partition') do |p|
-        p.elements('daemon') do |daemon|
-          parse_daemon_infos(daemon)
-        end
+      ['blocked', 'waiting'].each do |state|
+        wdata.search("//#{state}").each { |e| save_value_line(e, state) }
       end
+      wdata.search('partition/daemon').each { |daemon| parse_daemon_infos(daemon) }
       parse_worker_infos(wdata)
-      @squeues.each_pair do |key, value|
-        StatusHistory.create(time: @mytime, key: key, value: value)
-      end
+      @squeues.each_pair { |key, value| StatusHistory.create(time: @mytime, key: key, value: value) }
     end
-    ret
+
+    backend_status
   end
 
   private
 
   def add_squeue(key, value)
-    @squeues[key] ||= 0
     @squeues[key] += value.to_i
   end
 
   def parse_daemon_infos(daemon)
-    return unless daemon['type'] == 'scheduler'
-    arch = daemon['arch']
-    # FIXME2.5: The current architecture model is a gross hack, not connected at all
```

Comment:
And where do you create architectures now?

Reply:
Looking in master we don't create it here https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/worker_status.rb#L79 @coolo, and probably we won't create `Achitecture` while calculating the statistics, but I understand that I made it dependent of an existing `Architecture`, but mainly because my assumption is if there is a worker for a specific architecture, this architecture should exist in the DB, already, no?

Reply:
But it's set to available here

Reply:
here too `Architecture.find_by_name(arch).tap { |a| a.update(available: true) }`

Reply:
@coolo did the last answer, answered your question about making the Architecture available?

Reply:
Well, yes - but now you find and update the architecture in a loop instead of once. Doesn't appear ideal to me.

Reply:
> Well, yes - but now you find and update the architecture in a loop instead of once. Doesn't appear ideal to me.

ah, yes well, Rails provides a SQL query cache, used to cache the results of database queries for the duration of the request. So while we are in the same request we are good: https://guides.rubyonrails.org/caching_with_rails.html#sql-caching

Reply:
I changed that little bit @coolo. I added a scope for `Architecture` named `unavailable` and just update the objects if necessary.. and I do everything outside the loop, to avoid confusion.